pkg/connection/reconcile_topic.go

```diff
@@ -144,6 +154,38 @@ func (r *PulsarTopicReconciler) ReconcileTopic(ctx context.Context, pulsarAdmin
 		return err
 	}
 
+	schema, serr := pulsarAdmin.GetSchema(topic.Spec.Name)
+	if serr != nil && !admin.IsNotFound(serr) {
+		return serr
+	}
+	if topic.Spec.SchemaInfo != nil {
+		// Only upload the schema when schema doesn't exist or the schema has been updated
+		if admin.IsNotFound(serr) || !reflect.DeepEqual(topic.Spec.SchemaInfo, schema) {
+			info := topic.Spec.SchemaInfo
+			param := &admin.SchemaParams{
+				Type:       info.Type,
+				Schema:     info.Schema,
+				Properties: info.Properties,
+			}
+			log.Info("Upload schema for the topic", "name", topic.Spec.Name, "type", info.Type, "schema", info.Schema, "properties", info.Properties)
+			if err := pulsarAdmin.UploadSchema(topic.Spec.Name, param); err != nil {
+				log.Error(err, "Failed to upload schema")
+				if err := r.conn.client.Status().Update(ctx, topic); err != nil {
+					log.Error(err, "Failed to upload schema for the topic")
+					return nil
+				}
+				return err
+			}
+		}
+	} else if schema != nil && topic.Spec.SchemaInfo == nil {
```

Comment:
`topic.Spec.SchemaInfo == nil` is a redundant check here because of the `if` and I think you can move the lines 157 - 160 down past the `if` since you won't be using it `topic.Spec.SchemaInfo` is not nil.
e.g.

```
if topic.Spec.SchemaInfo != nil {
  ...
}

schema, serr := pulsarAdmin.GetSchema(topic.Spec.Name)
if serr != nil && !admin.IsNotFound(serr) {
  return serr
}

if schema != nil {
 ...
}

```

Reply:
yeah, line 180 is a redundant check. I will remove it.

Reply:
But for lines 157-160, If I move them down, I should add the check `topic.Spec.SchemaInfo == nil` back. Because what I want is to delete the existing schema when the `topic.spec.SchemaInfo` is updated to empty. If I only check the `schema != nil`, it will delete the existing schema no matter whether the `topic.spec.SchemaInfo` is empty or not. Or I just keep it at the `else if`.  e.g. 

```golang
if topic.Spec.SchemaInfo != nil {
  ...
}

schema, serr := pulsarAdmin.GetSchema(topic.Spec.Name)
if serr != nil && !admin.IsNotFound(serr) {
  return serr
}

if schema != nil && topic.Spec.SchemaInfo == nil {
  // Delete the schema when the schema exists and schema info is updated to empty.
 ...
}
```